draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1216 @@
+  should.
+
+  Too much customizability would break this promise.
+
+Click describes it's
+`advantages over argparse <https://click.palletsprojects.com/en/7.x/why/#why-not-argparse>`_
+in its documentation as well:
+
+  Click is internally based on optparse instead of argparse.  This however
+  is an implementation detail that a user does not have to be concerned
+  with.  The reason however Click is not using argparse is that it has some
+  problematic behaviors that make handling arbitrary command line interfaces
+  hard:
+
+  * argparse has built-in magic behavior to guess if something is an
+    argument or an option.  This becomes a problem when dealing with
+    incomplete command lines as it's not possible to know without having a
+    full understanding of the command line how the parser is going to
+    behave.  This goes against Click's ambitions of dispatching to
+    subparsers.
+  * argparse currently does not support disabling of interspersed
+    arguments.  Without this feature it's not possible to safely implement
+    Click's nested parsing nature.
+
+In contrast to :mod:`argparse`, the `Click community <https://github.com/click-contrib>`_
+provides many extensions we can use to create a better user experience
+for our users.
+
+Click supports calling `async` methods and functions
+using the `trio-click <https://github.com/click-contrib/trio-click>`_ extension
+which is likely to be important for us in the future.
+
+Worker
+------
+
+The Worker is the most fundamental architectural component in Celery.
+
+Event Loop
+++++++++++
+
+In Celery 4 we have implemented our own custom Event Loop.
+It is a cause for many bugs and issues in Celery.
+
+In addition, some I/O operations are still blocking the event loop since
+the clients we use do not allow non-blocking operations.
+
+The most important feature of Celery 5 is to replace the custom Event Loop
+with `Trio`_.
+
+We selected it because of it's
+`design <https://trio.readthedocs.io/en/latest/design.html>`_,
+`interoperability with asyncio <https://github.com/python-trio/trio-asyncio>`_
+and it's many features.
+
+Trio provides a context manager which limits the concurrency of coroutines
+and/or threads. This saves us from further bookkeeping when a Worker executes
+:ref:`draft/celery-5-high-level-architecture:Tasks`.
+
+Trio allows coroutines to report their status. This is especially useful
+when we want to block the execution of other coroutines until initialization
+of the coroutine completes. We require this feature for implementing
+:ref:`draft/celery-5-high-level-architecture:Boot Steps`.
+
+Trio also provides a feature called cancellation scopes which allows us to
+cancel a coroutine or multiple coroutines at once.
+This allows us to abort :ref:`draft/celery-5-high-level-architecture:Tasks`
+and handle the aborted tasks in an idiomatic fashion.
+
+All of those features save us from writing a lot of code. If we were to select
```

Comment:
Perhaps I am missing the point here, and haven't seen `trio` APIs in detail yet, but don't we also need multiprocessing as a choice for non-thread safe code or processing CPU-bound tasks? So my guess would be that some common abstraction layer would be also required.

Reply:
No, we do not.
trio can run subprocesses asynchronously if needed.

I'd rather make threads our default though as forking has created a lot of issues in the past.
I'll sum up the pros and cons in the document.

Reply:
I added some information.